### ISSUE TITLE
Not able to achieve published training performance

### ISSUE THREAD

**Issue by lvermue** [uma-pi1/kge#161]
Hi LibKGE-Team,
Amazed by the results you show in Table 2 from the [paper](https://www.aclweb.org/anthology/2020.emnlp-demos.22.pdf) for this library, I've been trying to understand how you get it that fast.

To start, I have installed the latest version of this repository.
For the test I created following script, this particular example for 1024 negative samples:
```
job.type: train
dataset.name: fb15k
model: complex

train:
  type: negative_sampling
  optimizer: Adagrad
  optimizer_args:
    lr: 0.6
  loss: bce
  loss_arg: 0.0
  batch_size: 512
  num_workers: 4

lookup_embedder.dim: 128

negative_sampling:
  num_samples:
    s: 512
    o: 512 
```

I invoke the training by running
`kge start examples/my-example-above.yaml`

I run this on a Titan XP GPU with CUDA 10.2, and an Intel Xeon E5-2620 v4.0 CPU, dataset on SSD.
It's not the exact hardware, but similar, yet it takes about ~65 seconds per epoch for me instead of the ~14 seconds I was expecting.
Also when I try to run 16384 negative samples, I get CUDA OOM errors.
Did I miss something in the configuration?

**Comment by AdrianKs** [comment]
Hi,
the trick is to set:
--negative_sampling.implementation batch

you can find the relevant documentation here:
https://github.com/uma-pi1/kge/blob/master/kge/config-default.yaml#L368

```yaml
  # 'batch' or 'auto' is recommended (faster) for models which have efficient
  # implementations to score many targets at once. For all other models, use
  # 'triple' (e.g., for TransE or RotatE in the current implementation).
  implementation: triple
```

This should also solve the OOM issue with 16384 negatives.

Please let me know if you can achieve similar epoch times with this setting.

**Comment by lvermue** [comment]
@AdrianKs Thanks, that made it definitely faster (~20 seconds now) 😊
I have a question regarding the scores that are generated by the batch processing.
When I use the above configuration the resulting `scores` tensor always has the shape [512, 513], the first column being the scores of the true triples and the others the scores of the negative samples.
However, it does this both for S and P indicating that the loss function takes into account the scores of the true triples twice for S and P. Is that correct?
https://github.com/uma-pi1/kge/blob/0347d4a36104a0adf6415469f07b0b4e121d28f9/kge/job/train.py#L1027

**Comment by rgemulla** [comment]
Yes, each slot is treated separately. You may think of one LibKGE epoch roughly as one epoch per slot. (Some other implementations actually only score one slot per batch.)

**Comment by lvermue** [comment]
But doesn't it over-sample the true triples then?
In this case I have `batch_size = 512`, so I can only have 512 true triples in that batch that are scored and back-propagated once, but in this case it happens twice.

**Comment by rgemulla** [comment]
For each slot, the true triple is scored precisely once. It's not over-sampling, it's training "per slot" instead of "across slots". The overall loss is the sum of the per-slot losses. It's consistently done like this in LibKGE.

(Note that there can be more than 512 true triples in a batch as well when `KvsAll` training is used.)

**Comment by samuelbroscheit** [comment]
@lvermue I am closing this issue as the original question was resolved. 

Please feel free to discuss other topics in a new issue. The design choice from our side w.r.t. "per slot" vs "across slots" was that we wanted a method that works for all training strategies: `negative_sampling`, `1vsAll` and `KvsAll`. The blocker for "across slots" is `KvsAll`.
